### PR TITLE
fix: upload artifacts in release process

### PR DESCRIPTION
**Pull request by freedmand**
## Description

The current release process is buggy; it doesn't actually upload release artifacts. This seems to be because I'm not passing files appropriately between two different jobs. An easier solution is to refine the reusable workflow jobs to handle multiple steps, which preserves access to all the files. This PR refines the release and Python wheel reusable workflows to accept these steps, simplifying and hopefully fixing the release process.

### CODE REVIEW

File: .github/workflows/_release.yml [hunk 1]
```diff
@@ -42,3 +46,37 @@ jobs:
...
+            libfastfec-linux-aarch64-${{ inputs.version }}.so
+            fastfec-macos-x86_64-${{ inputs.version }}.zip
+            libfastfec-macos-x86_64-${{ inputs.version }}.dylib
+            fastfec-macos-aarch64-${{ inputs.version }}.zip
+            libfastfec-macos-aarch64-${{ inputs.version }}.dylib
+            fastfec-windows-x86_64-${{ inputs.version }}.zip
+            libfastfec-windows-x86_64-${{ inputs.version }}.dll
+            fastfec-windows-aarch64-${{ inputs.version }}.zip
+            libfastfec-windows-aarch64-${{ inputs.version }}.dll
+            libfastfec-${{ inputs.version }}.wasm
```
Comment (chriszs) [inline .github/workflows/_release.yml:82]: This is fine, but you might be able to cut down on repetition here and above in the build step by matrixing on the target and artifact file names.

**Review by chriszs — DISMISSED** [review]
Looks good. One optional extra.

File: .github/workflows/_release.yml (hunk 1 above)
Comment (freedmand) [reply .github/workflows/_release.yml]: Ok, trying out the matrix strategy. Should rapidly speed up the GH actions too

File: .github/workflows/_release.yml [hunk 2]
```diff
@@ -23,60 +29,71 @@ jobs:
...
+              "windows": {
+                "ext": "dll",
+                "name": "fastfec"
+              },
+              "wasm": {
+                "ext": "wasm",
+                "name": "fastfec"
+              }
+            }
+          export_to: output
```
Comment (chriszs) [inline .github/workflows/_release.yml:57]: Ah, neat! Could maybe have done this natively with include? Or did using include for wasm preclude that?

**Comment by freedmand** [comment]
Ok @chriszs, ready for your re-review. Overview of steps taken:
* refactored the build/release steps to use a matrix. Wasm required some extra care but thankfully there's an include flag to inject an item into the matrix
* refactor the Python wheels step to work in parallel (with a slight modification to the python script to build wheels to support passing in a specific target)
* refactor the test step to also do two things in parallel: test C code and test Python code

As a result, it seems like all the actions finished in 1m 36s, which includes building 6 executables, 7 libraries, 6 Python wheels, and testing!!!

File: .github/workflows/_release.yml [hunk 3]
```diff
@@ -23,60 +29,71 @@ jobs:
...
+      - name: Move output library
+        if: matrix.os != 'wasm'
+        run: mv zig-out/lib/${{ steps.library.outputs.name }}.${{ steps.library.outputs.ext }} libfastfec-${{ matrix.os }}-${{ matrix.architecture }}-${{ inputs.version }}.${{ steps.library.outputs.ext }}
+      - name: Move output library
+        if: matrix.os == 'wasm'
```
Comment (chriszs) [inline .github/workflows/_release.yml:71]: Wasm sure seems like the dividing line between a lot of these and the source of some repition. I wonder if there's a way to consolidate either by forming the file names earlier or branching around wasm.

**Review by chriszs — APPROVED** [review]
This looks great. Optional thoughts appended.

File: .github/workflows/_release.yml (hunk 3 above)
Comment (freedmand) [reply .github/workflows/_release.yml:71]: Yea I think for now it's fine and still pretty minimal. There's for sure elegant ways to improve it that could be done, but it's so much better

File: .github/workflows/_release.yml (hunk 2 above)
Comment (freedmand) [reply .github/workflows/_release.yml:57]: I think the matrix doesn't allow you to map additional data per target. If I'm wrong, I'll open a new PR to fix that

Comment (chriszs) [reply .github/workflows/_release.yml:57]: If I'm reading the [include docs](https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs#expanding-or-adding-matrix-configurations) correctly, it should. But it's a confusing feature and I'm not 100% on everything it does.

**Pull request merged**